llama_cpp/llama.py

```diff
@@ -621,10 +603,52 @@ def set_seed(self, seed: int):
             seed: The random seed.
         """
         self._seed = seed
+    
+    def set_lora_adapter_scale(self, lora_path: str, scale: float, *, load_if_needed=False):
+        """
+        Set the scale for a LoRA adapter or 0.0 to disable it for inference. If the LoRA adapter file
+        has previously been loaded then this method will set its scale. If the LoRA adapter file has
+        not been previously loaded, this method will raise an exception, unless load_if_needed is set.
+        
+        Args:
+            lora_path: The path to the LoRA adapter. This path must have been loaded when the `Llama` object was created.
+            scale: The scaling factor to apply to the LoRA adapter. If 0.0, the LoRA adapter will be disabled so it won't be used during inference.
+            load_if_needed: Whether or not to load the adapter if it has not been previously been loaded. If True, this
+                method will attempt to load the adapter from the lora_path if needed. If False, loading an adapter that
+                hasn't already been loaded will raise an exception.
+        """
+        # Load adapter if needed (even if scale 0.0)
+        lora_adapter = self._lora_adapters_paths.get(lora_path)
+        if lora_adapter is None:
+            lora_adapter = internals.LlamaLoraAdapter(
+                self._model,
+                lora_path,
+                verbose=self.verbose,
+            )
+            if lora_adapter is None:
+                raise RuntimeError(
+                    f"Failed to initialize LoRA adapter from lora path: {lora_path}"
+                )
+            self._lora_adapters_paths[lora_path] = lora_adapter
+
+        if scale == 0.0:
+            # Remove from context; safe to call even if not in context
+            self._ctx.lora_adapter_remove(lora_adapter)
+        else:
+            # Set scale in context
+            self._ctx.lora_adapter_set(lora_adapter, scale)
+
+        if self.lora_adapters is None:
+            self.lora_adapters = {}
+        self.lora_adapters[lora_path] = scale
+        self._lora_adapters_active = tuple(sorted(
+            filter(lambda path_scale: path_scale[1] != 0.0, self.lora_adapters.items())
+        ))
```

Comment:
Cache a tuple with the active adapters - sorted and filtered so it can be used as a canonical cache key.